crates/amalthea/src/kernel.rs

```diff
@@ -419,17 +436,28 @@ fn zmq_forwarding_thread(
 
     // Forwards 0MQ message from the frontend to the corresponding
     // Amalthea channel.
-    let forward_inbound = || -> anyhow::Result<()> {
-        let msg = Message::read_from_socket(&stdin_socket);
-        stdin_inbound_tx.send(msg)?;
+    let forward_inbound =
+        |socket: &Socket, inbound_tx: &Sender<crate::Result<Message>>| -> anyhow::Result<()> {
+            let msg = Message::read_from_socket(socket);
+            inbound_tx.send(msg)?;
+            Ok(())
+        };
+
+    // Forwards special 0MQ XPUB subscription message from the frontend to the IOPub thread.
+    let forward_inbound_subscription = |socket: &Socket,
+                                        inbound_tx: &Sender<crate::Result<SubscriptionMessage>>|
+     -> anyhow::Result<()> {
+        let msg = SubscriptionMessage::read_from_socket(socket);
+        inbound_tx.send(msg)?;
         Ok(())
```

Comment:
We need a special inbound helper here, since this is a special `SubscriptionMessage` not a typical `Message`. It has its own format that it reads off the wire